File: Part 1 - Data Preprocessing/DataProcessing.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# Import libraries

import numpy as np  #  contains math tools
import matplotlib.pyplot as plt  # plot nice chart
import pandas as pd  # import dataset and manage dataset

# Import dataset
dataset = pd.read_csv('Data.csv')
#       first colon take all the line
#       :-1 take all the columns except the last one
#       .values take  all the values
#       => create matrix of independent variables
X = dataset.iloc[:, :-1].values  
#       create independent vector
y = dataset.iloc[:, 3].values

# Taking care of Mising data
#       sklearn: library to make machine model
#       preprocessing: contains classes/methods to prepocess data
#       import Imputer class: to take care of missing data
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
#       fit the imputer objct to matrix of feature X
imputer = imputer.fit(X[:, 1:3])
#       transform: replace missing data with missing value
X[:, 1:3] = imputer.transform(X[:, 1:3])

# Encoding Categorial data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
#       fit labelencoder_X to the first colimn Country of our matrix X
#       return the first column encoded
labelencoder_X = LabelEncoder()
X[:, 0] = labelencoder_X.fit_transform(X[:, 0])
#       Dummy variables
onehotencoder = OneHotEncoder(categorical_features = [0])
X = onehotencoder.fit_transform(X).toarray()

labelencoder_y = LabelEncoder()
y = labelencoder_y.fit_transform(y)

# Splitting dataset into Trainings Set and Test set
# X_train: training part of the matrix X
# X_test: testing part 
# y_train: training part of dependent variable that associated to X_train
# y_test: testing part of dependent vector  
# test_size = 0.5: half of data go to train, half go to test
# 0.2: 2 observations in test set, 8 in training set  
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test =  train_test_split(X, y, test_size=0.2, random_state = 0)

# Feature Scaling
from sklearn.preprocessing import StandardScaler 
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

File: Part 1 - Data Preprocessing/DataProcessing.R
# Data Processing

# Importing the dataset
dataset = read.csv('Data.csv')
#   take subset of dataset
# dataset = dataset[, 2:3]

# Taking care of missing data
#   take column Age of the dataset
#   ifelse: condition, return if True, return if False
#   is.na: function tells if value is missing or not
dataset$Age = ifelse(is.na(dataset$Age),
                     ave(dataset$Age, FUN = function(x) mean(x, na.rm = TRUE)),
                     dataset$Age)

dataset$Salary = ifelse(is.na(dataset$Salary),
                     ave(dataset$Salary, FUN = function(x) mean(x, na.rm = TRUE)),
                     dataset$Salary)

# Encoding categorial data
#   factor: function that 
#   first argument: data that want to tranforms into vector
#   second arguemnt: levels: name of categories in Country column
#   c: create vector of 3 elements
#   third argument: label: choose which number give to F, S, G
dataset$Country = factor(dataset$Country,
                         levels = c('France', 'Spain', 'Germany'),
                         labels = c(1, 2, 3))
dataset$Purchased = factor(dataset$Purchased,
                         levels = c('No', 'Yes'),
                         labels = c(0, 1))

# Splitting the dataset into Training set and Test set
# install.packages('caTools')
# .split: return True or False fro each of your observation:
# True if go to trainign set, False if go to Test Set
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling
training_set[, 2:3] = scale(training_set[, 2:3])
test_set[, 2:3] = scale(test_set[, 2:3])






